text/2017-09-04-generalized-scope.md

```diff
@@ -0,0 +1,125 @@
+# Summary
+
+This is a proposal for generalizing the `Scope` struct. It primarily aims to optimize the
+`unprotected` function. As a side benefit, we can easily support multiple instances of EBR-managed
+region of shared memory.
+
+
+# Motivation
+
+The `unprotected` function provides a `Scope` without pinning the thread, where the caller thread is
+the only one accessing atomics. Assuming this, we can **immediately** deallocate and drop objects in
+an unprotected scope. However, even in an unprotected scope, the `defer_free` and `defer_drop`
+functions defers freeing and dropping the objects until the scope is over. Even worse, if a lot of
+pieces of garbage are created, they are moved to the global garbage queue.
+
+Another motivation is the fact that currently we are supporting only a single instance of
+EBR-managed region of shared memory. For example, a bad-behaving mutator can pin itself
+indefinitely so that the global epoch cannot be advanced too long. In order to mitigate this
+problem, we would like to support multiple instances of EBR so that a mutator in an instance cannot
+bother the advancement of another instance's epoch.
+
+
+# Detailed Design
+
+This proposal is fully implemented
+in [this branch](https://github.com/jeehoonkang/crossbeam-epoch/tree/unprotected).
+
+
+## Generalizing `Scope`
+
+In order to enhance the `unprotected` function, we propose:
+
+- Creating
+  the
+  [`Scope` trait](https://github.com/jeehoonkang/crossbeam-epoch/blob/unprotected/src/realm.rs#L15),
```

Comment:
Instead of creating a split between `EpochScope` and `UnprotectedScope`, why not just put `is_unprotected: bool` inside `Scope`?